src/app/Library/CrudPanel/Traits/Validation.php

```diff
@@ -367,6 +250,77 @@ private function setupFieldValidation($field, $parent = false)
         }
     }
 
+    /**
+     * Return the messages for the fields and subfields in the current crud panel.
+     *
+     * @param  array  $fields
+     * @return array
+     */
+    private function getValidationMessagesFromFieldsAndSubfields($fields)
+    {
+        $messages = [];
+        collect($fields)
+            ->filter(function ($value, $key) {
+                // only keep fields where 'validationMessages' OR there are subfields
+                return array_key_exists('validationMessages', $value) || array_key_exists('subfields', $value);
+            })->each(function ($item, $key) use (&$messages) {
+                if (isset($item['validationMessages'])) {
+                    foreach ($item['validationMessages'] as $rule => $message) {
+                        $messages[$key.'.'.$rule] = $message;
+                    }
+                }
+                // add messages from subfields
+                if (array_key_exists('subfields', $item)) {
+                    $subfieldsWithValidationMessages = array_filter($item['subfields'], function ($subfield) {
+                        return array_key_exists('validationRules', $subfield);
+                    });
+
+                    foreach ($subfieldsWithValidationMessages as $subfield) {
+                        foreach ($subfield['validationMessages'] ?? [] as $rule => $message) {
+                            $messages[$item['name'].'.*.'.$subfield['name'].'.'.$rule] = $message;
+                        }
+                    }
+                }
+            })->toArray();
+
+        return $messages;
+    }
+
+    /**
+     * Return the rules for the fields and subfields in the current crud panel.
+     *
+     * @param  array  $fields
+     * @return array
+     */
+    private function getValidationRulesFromFieldsAndSubfields($fields)
```

Comment:
Note to self: this hasn't changed at all, just moved in a different place in the same file.